module-1/lab-functional-programming/your-code/Q2.ipynb

```diff
@@ -15,12 +15,139 @@
   },
```

Comment:
1. One possible way of doing the first part: 
```Python
# Define your string handling functions below
# Minimal 3 functions

import re
def strip_html_tags(html):
    s = re.sub("<!--.*?-->", "", html)
    s = re.sub("<.*?>", "", s)
    return(s)

from string import punctuation
def remove_punctuation(html):
    s = ''.join(c for c in html if c not in punctuation)
    return(s)

def to_lower_case(html):
    return(html.lower())
```
2. In the instruction " Next, paste your previously written get_bow_from_docs function below. Call your functions above at the appropriate place." you can do the following: (Take into account that this should be changed accordingly to your variables names. 
```Python
def get_bow_from_docs(docs, stop_words=[]):
    # In the function, first define the variables you will use such as `corpus`, `bag_of_words`, and `term_freq`.
    corpus = []
    bag_of_words = []
    term_freq = []
    
    """
    Loop `docs` and read the content of each doc into a string in `corpus`.
    Remember to convert the doc content to lowercases and remove punctuation.
    """
    for doc in docs:
        f = open(doc, 'r')
        s = f.read()
        s = strip_html_tags(s)
        s = remove_punctuation(s)
        s = to_lower_case(s)
        corpus.append(s)
        f.close()
    
    """
    Loop `corpus`. Append the terms in each doc into the `bag_of_words` array. The terms in `bag_of_words` 
    should be unique which means before adding each term you need to check if it's already added to the array.
    In addition, check if each term is in the `stop_words` array. Only append the term to `bag_of_words`
    if it is not a stop word.
    """
    for s in corpus:
        s = ''.join(c for c in s if c not in punctuation)
        terms = s.split()
        for term in terms:
            if not term in bag_of_words and not term in stop_words:
                bag_of_words.append(term)
    
    
    """
    Loop `corpus` again. For each doc string, count the number of occurrences of each term in `bag_of_words`. 
    Create an array for each doc's term frequency and append it to `term_freq`.
    """
    for s in corpus:
        freq = []
        terms = s.split()
        for word in bag_of_words:
            freq.append(terms.count(word))
        term_freq.append(freq)
    
    # Now return your output as an object
    return({
        "bag_of_words": bag_of_words,
        "term_freq": term_freq
    })
```